CAN-2004-0976:

  Multiple scripts in the perl package in Trustix Secure Linux 1.5 through
  2.1, and possibly other operating systems, allows local users to
  overwrite files via a symlink attack on temporary files.

 * some unsafe examples in the DB_File POD
 * use of the unsafe tmpnam in the ext/DB_File/t/db-recno.t test script,
 * use of unsafe temporary file names in ext/Devel/PPPort/PPPort.pm
 * An example in MakeMaker.pm that suggets setting PREFIX=/tmp/myperl5
   and another that suggets setting DESTDIR=/tmp/
 * Insecure use of /tmp file in instmodsh.
 * Insecure use of /tmp file in lib/Memoize/t/tie.t, tie_gdbm.t, tie_ndbm.t,
   tie_sdbm.t, tie_storable.t, probably exploitable at build time if these
   tests are run.
 * Use of insecure temp file in POD docs in lib/perl5db.pl
   and also an insecure temp file in the setterm() function in that
   program.

diff -Naur --exclude=debian perl-5.8.4.orig/ext/DB_File/DB_File.pm perl-5.8.4/ext/DB_File/DB_File.pm
--- perl-5.8.4.orig/ext/DB_File/DB_File.pm	2003-12-28 07:37:53.000000000 +1100
+++ perl-5.8.4/ext/DB_File/DB_File.pm	2005-02-02 22:53:02.000000000 +1100
@@ -1821,7 +1821,7 @@
     use DB_File ;
 
     my %hash ;
-    my $filename = "/tmp/filt" ;
+    my $filename = "filt" ;
     unlink $filename ;
 
     my $db = tie %hash, 'DB_File', $filename, O_CREAT|O_RDWR, 0666, $DB_HASH 
@@ -1863,7 +1863,7 @@
     use strict ;
     use DB_File ;
     my %hash ;
-    my $filename = "/tmp/filt" ;
+    my $filename = "filt" ;
     unlink $filename ;
 
 
@@ -1894,8 +1894,8 @@
 
 The locking technique went like this. 
 
-    $db = tie(%db, 'DB_File', '/tmp/foo.db', O_CREAT|O_RDWR, 0666)
-        || die "dbcreat /tmp/foo.db $!";
+    $db = tie(%db, 'DB_File', 'foo.db', O_CREAT|O_RDWR, 0666)
+        || die "dbcreat foo.db $!";
     $fd = $db->fd;
     open(DB_FH, "+<&=$fd") || die "dup $!";
     flock (DB_FH, LOCK_EX) || die "flock: $!";
diff -Naur --exclude=debian perl-5.8.4.orig/ext/DB_File/t/db-recno.t perl-5.8.4/ext/DB_File/t/db-recno.t
--- perl-5.8.4.orig/ext/DB_File/t/db-recno.t	2003-12-28 07:37:53.000000000 +1100
+++ perl-5.8.4/ext/DB_File/t/db-recno.t	2005-02-02 22:53:02.000000000 +1100
@@ -1198,7 +1198,7 @@
 
 my $testnum = 181;
 my $failed = 0;
-require POSIX; my $tmp = POSIX::tmpnam();
+my $tmp = "dbr$$";
 foreach my $test (@tests) {
     my $err = test_splice(@$test);
     if (defined $err) {
diff -Naur --exclude=debian perl-5.8.4.orig/ext/Devel/PPPort/PPPort.pm perl-5.8.4/ext/Devel/PPPort/PPPort.pm
--- perl-5.8.4.orig/ext/Devel/PPPort/PPPort.pm	2003-12-15 18:56:37.000000000 +1100
+++ perl-5.8.4/ext/Devel/PPPort/PPPort.pm	2005-02-02 22:53:02.000000000 +1100
@@ -349,13 +349,19 @@
 	}
 	
 	if ($changes) {
-		open(OUT,">/tmp/ppport.h.$$");
+		require POSIX; use Fcntl;
+		for(;;) {
+		    $tmp = POSIX::tmpnam();
+		    sysopen(OUT, $tmp, O_CREAT|O_WRONLY|O_EXCL, 0700) && last;
+		}
+
 		print OUT $c;
 		close(OUT);
-		open(DIFF, "diff -u $filename /tmp/ppport.h.$$|");
-		while (<DIFF>) { s!/tmp/ppport\.h\.$$!$filename.patched!; print STDOUT; }
+
+		open(DIFF, "diff -u $filename $tmp|");
+		while (<DIFF>) { s!$tmp!$filename.patched!; print STDOUT; }
 		close(DIFF);
-		unlink("/tmp/ppport.h.$$");
+		unlink($tmp);
 	} else {
 		print "Looks OK\n";
 	}
diff -Naur --exclude=debian perl-5.8.4.orig/lib/ExtUtils/MakeMaker.pm perl-5.8.4/lib/ExtUtils/MakeMaker.pm
--- perl-5.8.4.orig/lib/ExtUtils/MakeMaker.pm	2004-01-06 09:34:59.000000000 +1100
+++ perl-5.8.4/lib/ExtUtils/MakeMaker.pm	2005-02-02 22:53:02.000000000 +1100
@@ -1013,7 +1013,7 @@
 The Makefile to be produced may be altered by adding arguments of the
 form C<KEY=VALUE>. E.g.
 
-  perl Makefile.PL PREFIX=/tmp/myperl5
+  perl Makefile.PL PREFIX=~/myperl5
 
 Other interesting targets in the generated Makefile are
 
@@ -1355,13 +1355,13 @@
 
 This is the root directory into which the code will be installed.  It
 I<prepends itself to the normal prefix>.  For example, if your code
-would normally go into /usr/local/lib/perl you could set DESTDIR=/tmp/
-and installation would go into /tmp/usr/local/lib/perl.
+would normally go into /usr/local/lib/perl you could set DESTDIR=/other/
+and installation would go into /other/usr/local/lib/perl.
 
 This is primarily of use for people who repackage Perl modules.
 
 NOTE: Due to the nature of make, it is important that you put the trailing
-slash on your DESTDIR.  "/tmp/" not "/tmp".
+slash on your DESTDIR.  "/other/" not "/other".
 
 =item DIR
 
diff -Naur --exclude=debian perl-5.8.4.orig/lib/ExtUtils/instmodsh perl-5.8.4/lib/ExtUtils/instmodsh
--- perl-5.8.4.orig/lib/ExtUtils/instmodsh	2004-01-06 09:34:59.000000000 +1100
+++ perl-5.8.4/lib/ExtUtils/instmodsh	2005-02-02 22:53:02.000000000 +1100
@@ -2,6 +2,7 @@
 
 use strict;
 use IO::File;
+use File::Temp;
 use ExtUtils::Packlist;
 use ExtUtils::Installed;
 
@@ -58,16 +59,12 @@
       $reply =~ /^t\s*/ and do
          {
          my $file = (split(' ', $reply))[1];
-         my $tmp = "/tmp/inst.$$";
-         if (my $fh = IO::File->new($tmp, "w"))
-            {
-            $fh->print(join("\n", $Inst->files($module)));
-            $fh->close();
-            system("tar cvf $file -I $tmp");
-            unlink($tmp);
-            last CASE;
-            }
-         else { print("Can't open $file: $!\n"); }
+	 my ($fh, $tmp) = File::Temp::tempfile(UNLINK => 1);
+	 $fh->print(join("\n", $Inst->files($module)));
+	 $fh->close();
+	 # This used to use -I which is wrong for GNU tar.
+	 system("tar cvf $file -T $tmp");
+	 unlink($tmp);
          last CASE;
          };
       $reply eq 'v' and do
diff -Naur --exclude=debian perl-5.8.4.orig/lib/Memoize/t/tie.t perl-5.8.4/lib/Memoize/t/tie.t
--- perl-5.8.4.orig/lib/Memoize/t/tie.t	2002-07-13 05:56:19.000000000 +1000
+++ perl-5.8.4/lib/Memoize/t/tie.t	2005-02-02 22:53:02.000000000 +1100
@@ -29,14 +29,7 @@
   $_[0]+1;
 }
 
-if (eval {require File::Spec::Functions}) {
-  File::Spec::Functions->import('tmpdir', 'catfile');
-  $tmpdir = tmpdir();
-} else {
-  *catfile = sub { join '/', @_ };
-  $tmpdir = $ENV{TMP} || $ENV{TMPDIR} || '/tmp';
-}
-$file = catfile($tmpdir, "md$$");
+$file = "md$$";
 @files = ($file, "$file.db", "$file.dir", "$file.pag");
 1 while unlink @files;
 
diff -Naur --exclude=debian perl-5.8.4.orig/lib/Memoize/t/tie_gdbm.t perl-5.8.4/lib/Memoize/t/tie_gdbm.t
--- perl-5.8.4.orig/lib/Memoize/t/tie_gdbm.t	2002-07-13 05:56:19.000000000 +1000
+++ perl-5.8.4/lib/Memoize/t/tie_gdbm.t	2005-02-02 22:53:02.000000000 +1100
@@ -26,13 +26,7 @@
 
 print "1..4\n";
 
-if (eval {require File::Spec::Functions}) {
- File::Spec::Functions->import();
-} else {
-  *catfile = sub { join '/', @_ };
-}
-$tmpdir = $ENV{TMP} || $ENV{TMPDIR} ||  '/tmp';  
-$file = catfile($tmpdir, "md$$");
+$file = "md$$";
 1 while unlink $file, "$file.dir", "$file.pag";
 tryout('GDBM_File', $file, 1);  # Test 1..4
 1 while unlink $file, "$file.dir", "$file.pag";
diff -Naur --exclude=debian perl-5.8.4.orig/lib/Memoize/t/tie_ndbm.t perl-5.8.4/lib/Memoize/t/tie_ndbm.t
--- perl-5.8.4.orig/lib/Memoize/t/tie_ndbm.t	2002-07-13 05:56:19.000000000 +1000
+++ perl-5.8.4/lib/Memoize/t/tie_ndbm.t	2005-02-02 22:53:02.000000000 +1100
@@ -28,14 +28,7 @@
 
 print "1..4\n";
 
-
-if (eval {require File::Spec::Functions}) {
- File::Spec::Functions->import();
-} else {
-  *catfile = sub { join '/', @_ };
-}
-$tmpdir = $ENV{TMP} || $ENV{TMPDIR} ||  '/tmp';  
-$file = catfile($tmpdir, "md$$");
+$file = "md$$";
 1 while unlink $file, "$file.dir", "$file.pag";
 tryout('Memoize::NDBM_File', $file, 1);  # Test 1..4
 1 while unlink $file, "$file.dir", "$file.pag";
diff -Naur --exclude=debian perl-5.8.4.orig/lib/Memoize/t/tie_sdbm.t perl-5.8.4/lib/Memoize/t/tie_sdbm.t
--- perl-5.8.4.orig/lib/Memoize/t/tie_sdbm.t	2002-07-13 05:56:19.000000000 +1000
+++ perl-5.8.4/lib/Memoize/t/tie_sdbm.t	2005-02-02 22:53:02.000000000 +1100
@@ -28,14 +28,7 @@
 
 print "1..4\n";
 
-if (eval {require File::Spec::Functions}) {
- File::Spec::Functions->import('tmpdir', 'catfile');
- $tmpdir = tmpdir();
-} else {
- *catfile = sub { join '/', @_ };
-  $tmpdir = $ENV{TMP} || $ENV{TMPDIR} || '/tmp';
-}
-$file = catfile($tmpdir, "md$$");
+$file = "md$$";
 1 while unlink $file, "$file.dir", "$file.pag";
 tryout('Memoize::SDBM_File', $file, 1);  # Test 1..4
 1 while unlink $file, "$file.dir", "$file.pag";
diff -Naur --exclude=debian perl-5.8.4.orig/lib/Memoize/t/tie_storable.t perl-5.8.4/lib/Memoize/t/tie_storable.t
--- perl-5.8.4.orig/lib/Memoize/t/tie_storable.t	2002-07-13 05:56:19.000000000 +1000
+++ perl-5.8.4/lib/Memoize/t/tie_storable.t	2005-02-02 22:53:02.000000000 +1100
@@ -33,14 +33,7 @@
 
 print "1..4\n";
 
-
-if (eval {require File::Spec::Functions}) {
- File::Spec::Functions->import();
-} else {
-  *catfile = sub { join '/', @_ };
-}
-$tmpdir = $ENV{TMP} || $ENV{TMPDIR} ||  '/tmp';  
-$file = catfile($tmpdir, "storable$$");
+$file = "storable$$";
 1 while unlink $file;
 tryout('Memoize::Storable', $file, 1);  # Test 1..4
 1 while unlink $file;
diff -Naur --exclude=debian perl-5.8.4.orig/lib/perl5db.pl perl-5.8.4/lib/perl5db.pl
--- perl-5.8.4.orig/lib/perl5db.pl	2004-04-18 00:19:31.000000000 +1000
+++ perl-5.8.4/lib/perl5db.pl	2005-02-02 22:53:02.000000000 +1100
@@ -206,7 +206,7 @@
 =item * noTTY 
 
 if set, goes in NonStop mode.  On interrupt, if TTY is not set,
-uses the value of noTTY or "/tmp/perldbtty$$" to find TTY using
+uses the value of noTTY or "$HOME/.perldbtty$$" to find TTY using
 Term::Rendezvous.  Current variant is to have the name of TTY in this
 file.
 
@@ -5742,8 +5742,8 @@
         else {
             eval "require Term::Rendezvous;" or die;
             # See if we have anything to pass to Term::Rendezvous.
-            # Use /tmp/perldbtty$$ if not.
-            my $rv = $ENV{PERLDB_NOTTY} || "/tmp/perldbtty$$";
+            # Use $HOME/.perldbtty$$ if not.
+            my $rv = $ENV{PERLDB_NOTTY} || "$ENV{HOME}/.perldbtty$$";
 
             # Rendezvous and get the filehandles.
             my $term_rv = new Term::Rendezvous $rv;
diff -Naur --exclude=debian perl-5.8.4.orig/pod/perldebug.pod perl-5.8.4/pod/perldebug.pod
--- perl-5.8.4.orig/pod/perldebug.pod	2004-04-04 21:58:12.000000000 +1000
+++ perl-5.8.4/pod/perldebug.pod	2005-02-02 22:54:50.000000000 +1100
@@ -700,7 +700,7 @@
 with two methods: C<IN> and C<OUT>.  These should return filehandles to use
 for debugging input and output correspondingly.  The C<new> method should
 inspect an argument containing the value of C<$ENV{PERLDB_NOTTY}> at
-startup, or C<".perldbtty$$"> otherwise.  This file is not 
+startup, or C<"~/.perldbtty$$"> otherwise.  This file is not 
 inspected for proper ownership, so security hazards are theoretically
 possible.
 
diff -Naur --exclude=debian perl-5.8.4.orig/utils/c2ph.PL perl-5.8.4/utils/c2ph.PL
--- perl-5.8.4.orig/utils/c2ph.PL	2004-04-04 21:58:26.000000000 +1000
+++ perl-5.8.4/utils/c2ph.PL	2005-02-02 22:53:58.000000000 +1100
@@ -1320,7 +1320,7 @@
 	$intrinsics{$_[1]} = $template{$_[0]};
     }
     close(PIPE) || die "couldn't read intrinsics!";
-    unlink($TMP, '$SAFEDIR/a.out');
+    unlink($TMP, "$SAFEDIR/a.out");
     print STDERR "done\n" if $trace;
 }
 
